src/sage/rings/valuation/mapped_valuation.py

```diff
@@ -0,0 +1,562 @@
+        return self._base_valuation(self._to_base_domain(f))
+
+    def reduce(self, f):
+        r"""
+        Return the reduction of ``f`` in the :meth:`residue_field` of this valuation.
+
+        EXAMPLES::
+
+            sage: K.<x> = FunctionField(QQ)
+            sage: R.<y> = K[]
+            sage: L.<y> = K.extension(y^2 - (x - 2))
+
+            sage: v = K.valuation(0)
+            sage: w = v.extension(L)
+            sage: w.reduce(y)
+            u1
+
+        """
+        return self._base_valuation.reduce(self._to_base_domain(f))
+
+    def lift(self, F):
+        r"""
+        Lift ``F`` from the :meth;`residue_field` of this valuation into its
+        domain.
+
+        EXAMPLES::
+
+            sage: K.<x> = FunctionField(QQ)
+            sage: R.<y> = K[]
+            sage: L.<y> = K.extension(y^2 - x)
+
+            sage: v = K.valuation(2)
+            sage: w = v.extension(L)
+            sage: w.lift(w.residue_field().gen())
+            y
+
+        """
+        F = self.residue_ring().coerce(F)
+        F = self._to_base_residue_ring(F)
+        f = self._base_valuation.lift(F)
+        return self._from_base_domain(f)
+
+    def _to_base_residue_ring(self, F):
+        r"""
+        Return ``F``, an element of :meth:`residue_ring`, as an element of the
+        residue ring of the ``_base_valuation``.
+
+        EXAMPLES::
+
+            sage: K.<x> = FunctionField(QQ)
+            sage: R.<y> = K[]
+            sage: L.<y> = K.extension(y^2 - x)
+
+            sage: v = K.valuation(0)
+            sage: w = v.extensions(L)[0]
+            sage: w._to_base_residue_ring(1)
+            1
+
+        """
+        return self._base_valuation.residue_ring().coerce(F)
+
+    def _from_base_residue_ring(self, F):
+        r"""
+        Return ``F``, an element of the residue ring of ``_base_valuation``, as
+        an element of this valuation's :meth:`residue_ring`.
+
+        EXAMPLES::
+
+            sage: K.<x> = FunctionField(QQ)
+            sage: R.<y> = K[]
+            sage: L.<y> = K.extension(y^2 - x)
+
+            sage: v = K.valuation(0)
+            sage: w = v.extensions(L)[0]
+            sage: w._from_base_residue_ring(1)
+            1
+
+        """
+        return self.residue_ring().coerce(F)
+
+    def _test_to_from_base_domain(self, **options):
+        r"""
+        Check the correctness of :meth:`to_base_domain` and
+        :meth:`from_base_domain`.
+
+        EXAMPLES::
+
+            sage: K.<x> = FunctionField(QQ)
+            sage: R.<y> = K[]
+            sage: L.<y> = K.extension(y^2 - x)
+
+            sage: v = K.valuation(0)
+            sage: w = v.extensions(L)[0]
+            sage: w._test_to_from_base_domain()
+
+        """
+        tester = self._tester(**options)
+
+        for x in tester.some_elements(self.domain().some_elements()):
+            tester.assertEqual(x, self._from_base_domain(self._to_base_domain(x)))
+            # note that the converse might not be true
```

Comment:
Should there also be a tester for to_from_residue_ring?  Or add it to this test?